circus-train-s3-s3-copier/src/main/java/com/hotels/bdp/circustrain/s3s3copier/S3S3Copier.java

```diff
@@ -225,22 +226,74 @@ private void applyObjectMetadata(CopyObjectRequest copyObjectRequest) {
     }
   }
 
-  private Metrics gatherAllCopyResults() {
-    AtomicLong bytesReplicated = new AtomicLong(0);
-    registerRunningMetrics(bytesReplicated);
-    for (Copy copyJob : copyJobs) {
+  private void processAllCopyJobs() {
+    List<CopyJob> copyJobsToSubmit = copyJobs;
+    int maxCopyAttempts = s3s3CopierOptions.getMaxCopyAttempts();
+    for (int copyAttempt = 1; copyAttempt <= maxCopyAttempts; copyAttempt++) {
+      LOG
+          .info("Submitting {} copy job(s), attempt {}/{}", copyJobsToSubmit.size(), copyAttempt, maxCopyAttempts);
+      copyJobsToSubmit = submitAndGatherCopyJobs(copyJobsToSubmit);
+      if (copyJobsToSubmit.isEmpty()) {
+        LOG
+            .info("Successfully gathered all copy jobs on attempt {}/{}", copyAttempt, maxCopyAttempts);
+        break;
+      }
+      if (copyAttempt == maxCopyAttempts) {
+        throw new CircusTrainException(copyJobsToSubmit.size() + " job(s) failed the maximum number of copy attempts, " + maxCopyAttempts);
+      }
+      LOG
+          .info("Finished gathering jobs on attempt {}/{}. Retrying {} failed job(s).",
+              copyAttempt,
+              maxCopyAttempts,
+              copyJobsToSubmit.size());
+    }
+  }
+
+  private List<CopyJob> submitAndGatherCopyJobs(List<CopyJob> copyJobsToSubmit) {
+    List<CopyJob> submittedCopyJobs = new ArrayList<>();
+    for (CopyJob copyJob : copyJobsToSubmit) {
+      Copy copy = submitCopyJob(copyJob);
+      CopyJob newCopyJob = new CopyJob(copy, copyJob.getCopyObjectRequest(), copyJob.getTransferStateChangeListener());
+      submittedCopyJobs.add(newCopyJob);
+    }
+    return gatherCopyJobs(submittedCopyJobs);
+  }
+
+  private Copy submitCopyJob(CopyJob copyJob) {
+    CopyObjectRequest copyObjectRequest = copyJob.getCopyObjectRequest();
+    LOG
+        .info("Copying object from '{}/{}' to '{}/{}'", copyObjectRequest.getSourceBucketName(), copyObjectRequest.getSourceKey(),
+            copyObjectRequest.getDestinationBucketName(), copyObjectRequest.getDestinationKey());
+    return transferManager.copy(copyObjectRequest, srcClient, copyJob.getTransferStateChangeListener());
+  }
+
+  private List<CopyJob> gatherCopyJobs(List<CopyJob> copyJobs) {
+    List<CopyJob> failedCopyJobs = new ArrayList<>();
+    for (CopyJob copyJob : copyJobs) {
       try {
-        copyJob.waitForCompletion();
-        long alreadyReplicated = bytesReplicated.addAndGet(copyJob.getProgress().getTotalBytesToTransfer());
-        if (totalBytesToReplicate > 0) {
+        Copy copy = copyJob.getCopy();
+        try {
+          copy.waitForCompletion();
+          long alreadyReplicated = bytesReplicated.addAndGet(copy.getProgress().getTotalBytesToTransfer());
+          if (totalBytesToReplicate > 0) {
+            LOG
+                .info("Replicating...': {}% complete",
+                    String.format("%.0f", (alreadyReplicated / (double) totalBytesToReplicate) * 100.0));
+          }
+        } catch (AmazonClientException e) {
```

Comment:
We're not logging this anywhere now. Perhaps add a `LOG.debug("Copying failed, retrying, exception is", e); ` or something similar.

Reply:
Added an extra debug line to log the exception